test/integration/ctl/migrate/ctl_upgrade_migrate_test.go

```diff
@@ -0,0 +1,202 @@
+/*
+Copyright 2022 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package migrate
+
+import (
+	"context"
+	"os"
+	"testing"
+	"time"
+
+	testlogger "github.com/go-logr/logr/testing"
+	apiextinstall "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/install"
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/jetstack/cert-manager/cmd/ctl/pkg/upgrade/migrate"
+	"github.com/jetstack/cert-manager/pkg/webhook/handlers"
+	"github.com/jetstack/cert-manager/pkg/webhook/handlers/testdata/apis/testgroup/install"
+	"github.com/jetstack/cert-manager/pkg/webhook/handlers/testdata/apis/testgroup/v1"
+	"github.com/jetstack/cert-manager/pkg/webhook/handlers/testdata/apis/testgroup/v2"
+	"github.com/jetstack/cert-manager/test/integration/framework"
+)
+
+var (
+	equivalentResources = map[string]client.Object{
+		"v1": &v1.TestType{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "object",
+				Namespace: "default",
+			},
+			TestField:          "abc",
+			TestFieldImmutable: "def",
+		},
+		"v2": &v2.TestType{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "object",
+				Namespace: "default",
+			},
+			TestField:          "abc",
+			TestFieldImmutable: "def",
+		},
+	}
+)
+
+func newScheme() *runtime.Scheme {
+	scheme := runtime.NewScheme()
+	apiextinstall.Install(scheme)
+	install.Install(scheme)
+	return scheme
+}
+
+func TestCtlUpgradeMigrate(t *testing.T) {
```

Comment:
The test looks great 👍🏼 I've tried to break it by commenting out the bits that run the migrator and verified that in that case [the update step](https://github.com/jetstack/cert-manager/pull/4711/commits/5a7a413a798cce029965404141ab1c9790ec06a8#diff-6251668df35e183fc5f89fcad4d9e4759711275937800167612ec8470e0bd934R147) fails. I assume that the bit that tries to list resources would also be able to detect if the migrator wasn't actually updating resources, but not sure how to verify that

Reply:
Yep - because we persist the object initially with the storedVersion set to v2, and then switch the storage version to v1 without performing our own explicit write, this is verifying that:

1) objects are actually read and then written again (i.e. re-persisted into the new storage version)
2) the migrator is patching `status.storedVersion`

If the migrator didn't do its job properly, the 'List' would fail as the API server can no longer perform conversion. I can write another test similar to this that verifies that if we want - though it'd be a bit contrived as it'd ultimately be verifying Kubernetes behaviour so wouldn't call into any non-test code.

Reply:
Thanks for adding that explanation - this makes sense 👍🏼 

I think if we wanted to add more tests, then perhaps it might make sense to test the flag that forces all resources to be migrated and also that the command fails if the current storage version is not v1

Reply:
I've added both of these tests now 🎉